cmd/snap/cmd_snap_op.go

```diff
@@ -138,10 +138,68 @@ type cmdRemove struct {
 	} `positional-args:"yes" required:"yes"`
 }
 
+func showRemoved(expectedSnaps []string, expectedComps map[string][]string, snapToRev map[string]string, removed *changedSnapsData) {
+	showRemovedSnaps(expectedSnaps, removed.names, snapToRev)
+	showRemovedComponents(expectedComps, removed.comps)
+}
+
+func showRemovedSnaps(expectedSnaps []string, removedSnaps []string, snapToRev map[string]string) {
+	seen := make(map[string]bool)
+	for _, name := range removedSnaps {
+		if rev, ok := snapToRev[name]; ok {
+			fmt.Fprintf(Stdout, i18n.G("%s (revision %s) removed\n"), name, rev)
+		} else {
+			fmt.Fprintf(Stdout, i18n.G("%s removed\n"), name)
+		}
+		seen[name] = true
+	}
+	for _, name := range expectedSnaps {
+		if !seen[name] {
+			// FIXME: this is the only reason why a name can be
+			// skipped, but it does feel awkward
+			fmt.Fprintf(Stdout, i18n.G("%s not installed\n"), name)
+		}
+	}
+}
+
+func showRemovedComponents(expectedBySnap, removedBySnap map[string][]string) {
+	snapCompsSeen := make(map[string]bool, len(removedBySnap))
+	for snapName, removedComps := range removedBySnap {
+		snapCompsSeen[snapName] = true
+
+		for _, comp := range removedComps {
+			fmt.Fprintf(Stdout, i18n.G("component %s for %s removed\n"), comp, snapName)
+		}
+
+		expected := expectedBySnap[snapName]
+		for _, comp := range expected {
+			if strutil.ListContains(removedComps, comp) {
+				continue
+			}
+			// FIXME: this is the only reason why a name can be
+			// skipped, but it does feel awkward
+			fmt.Fprintf(Stdout, i18n.G("component %s for %s is not installed\n"),
+				comp, snapName)
+		}
+	}
+	// Case of no component of the expected ones removed for a snap
+	for snapWithComps, expected := range expectedBySnap {
+		if _, ok := snapCompsSeen[snapWithComps]; ok {
+			continue
+		}
+		for _, comp := range expected {
+			// See FIXME comment above
+			fmt.Fprintf(Stdout, i18n.G("component %s for %s is not installed\n"),
+				comp, snapWithComps)
+		}
+	}
+}
+
 func (x *cmdRemove) removeOne(opts *client.SnapOptions) error {
-	name := string(x.Positional.Snaps[0])
+	arg := string(x.Positional.Snaps[0])
 
-	changeID, err := x.client.Remove(name, opts)
+	name, comps := splitSnapAndComponents(arg)
+	changeID, err := x.client.Remove(name, comps, opts)
```

Comment:
if comps is not empty this remove only the components and not the snap right?

Reply:
Yes, that is right, I've added a comment to clarify.